docs/source/contribution/cpu_features.rst

```diff
@@ -0,0 +1,214 @@
+Those files contain the definitions of one or several template classes that define member functions that
+do the actual computations. Here is a variant of the ``Abc`` training algorithm kernel definition in the file
+`abc_classification_train_kernel.h`:
+
+.. include:: ../includes/cpu_features/abc-classification-train-kernel.rst
+
+Typical template parameters are:
+
+- ``algorithmFPType``  Data type to use in intermediate computations for the algorithm,
+                       ``float`` or ``double``.
+- ``method`` Computational methods of the algorithm. ``method1`` or ``method2`` in the case of ``Abc``.
+- ``cpu`` Version of the cpu-specific implementation of the algorithm, ``daal::CpuType``.
+
+Implementations for different methods are usually defined usind partial class templates specialization.
+
+\*_impl.i
+---------
+
+Those files contain the implementations of the computational functions defined in `*_kernel.h` files.
+Here is a variant of ``method1`` imlementation for ``Abc`` training algorithm that does not contain any
+instruction set specific code. The implementation is located in the file `abc_classification_train_method1_impl.i`:
+
+.. include:: ../includes/cpu_features/abc-classification-train-method1-impl.rst
+
+Although the implementation of the ``method1`` does not contain any instruction set specific code, it is
+expected that the developers leverage SIMD related macros available in oneDAL.
+For example, ``PRAGMA_IVDEP``, ``PRAGMA_VECTOR_ALWAYS``, ``PRAGMA_VECTOR_ALIGNED`` and others pragmas defined in
+`service_defines.h <https://github.com/oneapi-src/oneDAL/blob/main/cpp/daal/src/services/service_defines.h>`_.
+This will guide the compiler to generate more efficient code for the target architecture.
+
+Consider that the implementation of the ``method2`` for the same algorithm will be different and will contain
+AVX-512-specific code located in ``cpuSpecificCode`` function.
+Then the implementation of the ``method2`` in the file `abc_classification_train_method2_impl.i` will look like:
+
+.. include:: ../includes/cpu_features/abc-classification-train-method2-impl.rst
+
+CPU-specific code needs to be placed under compiler-specific and CPU-specific defines because it usually
+contains intrinsics that cannot be compiled on other architectures.
+
+\*_fpt_cpu.cpp
+--------------
+
+Those files contain the instantiations of the template classes defined in `*_kernel.h` files.
+The instatiation of the ``Abc`` training algorithm kernel for ``method1`` is located in the file
+`abc_classification_train_method1_batch_fpt_cpu.cpp`:
+
+.. include:: ../includes/cpu_features/abc-classification-train-method1-fpt-cpu.rst
+
+`_fpt_cpu.cpp` files are not compiled directly into object files. First, multiple copies of those files
+are made replacing the ``fpt`` and ``cpu`` parts of the file name as well as the corresponding ``DAAL_FPTYPE`` and
+``DAAL_CPU`` macros with the actual data type and CPU type values. Then the resulting files are compiled
```

Comment:
Perhaps it could mention here that SIMD code also needs to be wrapped into an `#ifdef` section with conditions like `DAAL_INTEL_CPP_COMPILER && (__CPUID__(DAAL_CPU) == __avx512__)`, in addition to having a templated version.